parsl/execution_provider/aws/aws.py

```diff
@@ -0,0 +1,308 @@
+            'FromPort': 0,
+            'ToPort': 65535,
+            'IpRanges': ip_ranges,
+        }, ]
+
+        sg.authorize_ingress(IpPermissions=inPerms)
+        sg.authorize_egress(IpPermissions=outPerms)
+        self.sg_id = sg.id
+        return sg
+
+    def config_route_table(self, vpc, internet_gateway):
+        route_table = vpc.create_route_table()
+        route_ig_ipv4 = route_table.create_route(
+            DestinationCidrBlock='0.0.0.0/0',
+            GatewayId=internet_gateway.internet_gateway_id)
+        return route_table
+
+    # vpc = create_vpc(ec2, client)
+    # print(vpc.id)
+
+    def scale_out(self, size):
+        for i in range(size*self.config['nodeGranularity']):
+            self.spin_up_instance()
+
+    def scale_in(self, size):
+        for i in range(size*self.config['nodeGranularity']):
+            self.shut_down_instance()
+
+    def spin_up_instance(self):
+        instance_type = self.config['instancetype']
+        subnet = self.sn_ids[0]
+        ami_id = self.config['AMIID']
+        total_instances = len(self.instances) + self.config['nodeGranularity']
+        if total_instances > self.config['maxNodes']:
+            print(
+                "You have requested more instances ({}) than your maxNodes ({}). Cannot Continue".format(
+                    total_instances,
+                    self.config['maxNodes']))
+            return -1
+        instance = self.ec2.create_instances(
+            InstanceType=instance_type,
+            ImageId=ami_id,
+            MinCount=1,
+            MaxCount=self.config['nodeGranularity'],
+            KeyName=self.config['AWSKeyName'],
+            SubnetId=subnet,
+            SecurityGroupIds=[self.sg_id],
+            UserData=WORKER_USERDATA)
+        self.instances.append(instance[0].id)
+        self.write_state_file()
+        return instance
+
+    def shut_down_instance(self, instances=None):
+        if instances:
+            term = self.client.terminate_instances(InstanceIds=instances)
+        else:
+            instance = self.instances.pop()
+            try:
+                self.instance_states.pop(instance)
+            except KeyError:
+                pass
+            term = self.client.terminate_instances(InstanceIds=[instance])
+        self.get_instance_state()
+        return term
+
+    def get_instance_state(self, instances=None):
+        if instances:
+            desc = self.client.describe_instances(InstanceIds=instances)
+        else:
+            desc = self.client.describe_instances(InstanceIds=self.instances)
+        # pprint.pprint(desc['Reservations'],indent=4)
+        for i in range(len(desc['Reservations'])):
+            instance = desc['Reservations'][i]['Instances'][0]
+            self.instance_states[instance['InstanceId']
+                                 ] = instance['State']['Name']
+        self.write_state_file()
+        return self.instance_states
+
+    def submit(self):
```

Comment:
Submit, status, and cancel should follow a tighter pattern -> https://github.com/Parsl/parsl/blob/master/parsl/execution_provider/execution_provider_base.py